src/BSON/Int64.c

```diff
@@ -244,6 +263,229 @@ static int php_phongo_int64_compare_objects(zval* o1, zval* o2)
+ * prohibits us from falling back to PHP's default functions after unwrapping
+ * the int64_t from the php_phongo_int64_t instance. */
+static int64_t phongo_pow_int64(int64_t base, int64_t exp)
+{
+	if (exp == 0) {
+		return 1;
+	}
+
+	if (exp % 2) {
+		return base * phongo_pow_int64(base * base, (exp - 1) / 2);
+	}
+
+	return phongo_pow_int64(base * base, exp / 2);
+}
+
+#define OPERATION_RESULT_INT64(zval, value)                                          \
+	if (Z_TYPE_P((zval)) == IS_OBJECT && Z_OBJCE_P((zval)) == php_phongo_int64_ce) { \
+		Z_INT64_OBJ_P((zval))->integer = (value);                                    \
+	} else {                                                                         \
+		ZVAL_INT64_OBJ((zval), value);                                               \
+	}
+
+#define PHONGO_GET_INT64(int64, zval)                                                       \
+	if (Z_TYPE_P((zval)) == IS_LONG) {                                                      \
+		(int64) = Z_LVAL_P((zval));                                                         \
+	} else if (Z_TYPE_P((zval)) == IS_OBJECT && Z_OBJCE_P((zval)) == php_phongo_int64_ce) { \
+		(int64) = Z_INT64_OBJ_P((zval))->integer;                                           \
+	} else {                                                                                \
+		return FAILURE;                                                                     \
+	}
+
+/* Overload arithmetic operators for computation on int64_t values.
+ * This ensures that any computation involving at least one php_phongo_int64_t
+ * results in a php_phongo_int64_t value, regardless of whether the result
+ * would fit in an int32_t or not.
+ * Note that ZEND_(PRE|POST)_(INC|DEC) are not handled here: when checking for
+ * a do_operation handler for inc/dec, PHP calls the handler with a ZEND_ADD
+ * or ZEND_SUB opcode and the same pointer for result and op1, and a ZVAL_LONG
+ * of 1 for op2. This means that we can neither specially handle inc/dec
+ * operators, nor can we ensure correct behaviour for pre/post operators,
+ * leading to a difference in behaviour compared to int values. */
+static zend_result php_phongo_int64_do_operation(zend_uchar opcode, zval* result, zval* op1, zval* op2)
+{
+	int64_t value1, value2, lresult;
+
+	PHONGO_GET_INT64(value1, op1);
+
+	switch (opcode) {
+		case ZEND_ADD:
+			PHONGO_GET_INT64(value2, op2);
+
+			lresult = value1 + value2;
+
+			// Check for overflow by comparing the sign bits of the operands and the result
+			if ((value1 & INT64_MIN) == (value2 & INT64_MIN) && (value1 & INT64_MIN) != (lresult & INT64_MIN)) {
+				ZVAL_DOUBLE(result, (double) value1 + (double) value2);
+			} else {
+				OPERATION_RESULT_INT64(result, lresult);
+			}
+
+			return SUCCESS;
+
+		case ZEND_SUB:
+			PHONGO_GET_INT64(value2, op2);
+
+			lresult = value1 - value2;
+
+			// Check for overflow by comparing the sign bits of the operands and the result
```

Comment:
If you add a reference to `fast_long_add_function()` above, you can also reference `fast_long_sub_function()` here. I see the conditional logic is slightly different.

Just working through this truth table in my head, the two overflow cases (i.e. use a double) are:

 * Subtracting a negative from a positive (i.e. addition) yields a negative
 * Subtracting a positive from a negative yields a positive

That seems short enough to capture in a comment to help future readers.